File: frontend/src/types/purchase.ts
import type { Product } from './product';
import type { Supplier } from './supplier';

export type { Product };
export type { Supplier };

export type PaymentStatus = 'PAID' | 'PARTIAL' | 'UNPAID';

export type PurchaseStatus = 'DRAFT' | 'PENDING' | 'COMPLETED' | 'CANCELLED';

export type PurchaseTransactionType = 'PURCHASE' | 'PURCHASE_CHALLAN';

export type GstType = 'GST' | 'NON_GST';
export type TaxMode = 'INTRA_STATE' | 'INTER_STATE';
export type PaymentMode = 'CASH' | 'BANK' | 'UPI' | 'CARD' | 'OTHER';

export interface PurchaseItem {
  id: number;
  product: Product;
  quantity: number;
  purchasePrice: number;
  discount: number;
  tax: number;
  subtotal: number;
  gstPercent?: number;
  taxableAmount?: number;
  cgstAmount?: number;
  sgstAmount?: number;
  igstAmount?: number;
  purchaseOrderItemId?: number | null;
}

export interface Purchase {
  id: number;
  purchaseNumber: string;
  supplier: Supplier;
  purchaseDate: string;
  items: PurchaseItem[];
  subtotalAmount: number;
  totalDiscount: number;
  totalTax: number;
  totalAmount: number;
  paymentStatus: PaymentStatus;
  status: PurchaseStatus;
  notes?: string;
  gstType: GstType;
  taxMode?: TaxMode | null;
  supplierPhone?: string;
  supplierGstin?: string;
  billingAddress?: string;
  shippingAddress?: string;
  paymentMode: PaymentMode;
  paidAmount: number;
  payableAmount: number;
  taxableAmount: number;
  cgstAmount: number;
  sgstAmount: number;
  igstAmount: number;
  purchaseOrderId?: number | null;
  purchaseOrderNumber?: string | null;
  hasPayments: boolean;
  transactionType: PurchaseTransactionType;
  gstReportingApplicable: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface PurchaseItemPayload {
  productId: number;
  quantity: number;
  purchasePrice: number;
  discount: number;
  tax: number;
  gstPercent?: number;
  purchaseOrderItemId?: number;
}

export interface PurchaseCreatePayload {
  supplierId: number;
  purchaseDate: string;
  gstType: GstType;
  taxMode?: TaxMode | '';
  supplierPhone?: string;
  supplierGstin?: string;
  billingAddress?: string;
  shippingAddress?: string;
  paymentMode: PaymentMode;
  paidAmount: number;
  notes?: string;
  purchaseOrderId?: number;
  items: PurchaseItemPayload[];
  transactionType?: PurchaseTransactionType;
  saveAsDraft?: boolean;
}

export interface PurchaseUpdatePayload extends PurchaseCreatePayload {
  status: PurchaseStatus;
}

File: backend/src/main/java/com/storehub/entity/Purchase.java
package com.storehub.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "purchases")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Purchase {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "purchase_number", unique = true, length = 30)
    private String purchaseNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "supplier_id", nullable = false)
    private Supplier supplier;

    @Column(name = "purchase_date", nullable = false)
    private LocalDate purchaseDate;

    @Column(name = "total_amount", nullable = false, precision = 12, scale = 2)
    private BigDecimal totalAmount;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_status", nullable = false, length = 20)
    private PaymentStatus paymentStatus;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private PurchaseStatus status;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Enumerated(EnumType.STRING)
    @Column(name = "gst_type", length = 20)
    private GstType gstType;

    @Enumerated(EnumType.STRING)
    @Column(name = "tax_mode", length = 20)
    private TaxMode taxMode;

    @Column(name = "supplier_phone", length = 20)
    private String supplierPhone;

    @Column(name = "supplier_gstin", length = 20)
    private String supplierGstin;

    @Column(name = "billing_address", columnDefinition = "TEXT")
    private String billingAddress;

    @Column(name = "shipping_address", columnDefinition = "TEXT")
    private String shippingAddress;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_mode", length = 20)
    private PaymentMode paymentMode;

    @Column(name = "paid_amount", precision = 12, scale = 2)
    private BigDecimal paidAmount;

    @Column(name = "payable_amount", precision = 12, scale = 2)
    private BigDecimal payableAmount;

    @Column(name = "taxable_amount", precision = 12, scale = 2)
    private BigDecimal taxableAmount;

    @Column(name = "cgst_amount", precision = 12, scale = 2)
    private BigDecimal cgstAmount;

    @Column(name = "sgst_amount", precision = 12, scale = 2)
    private BigDecimal sgstAmount;

    @Column(name = "igst_amount", precision = 12, scale = 2)
    private BigDecimal igstAmount;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "purchase_order_id")
    private PurchaseOrder purchaseOrder;

    @Builder.Default
    @OneToMany(mappedBy = "purchase", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<PurchaseItem> items = new ArrayList<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
        if (this.status == null) {
            this.status = PurchaseStatus.PENDING;
        }
        if (this.totalAmount == null) {
            this.totalAmount = BigDecimal.ZERO;
        }
        if (this.gstType == null) {
            this.gstType = GstType.NON_GST;
        }
        if (this.paidAmount == null) {
            this.paidAmount = BigDecimal.ZERO;
        }
        if (this.payableAmount == null) {
            this.payableAmount = this.totalAmount;
        }
        if (this.taxableAmount == null) {
            this.taxableAmount = this.totalAmount;
        }
        if (this.cgstAmount == null) {
            this.cgstAmount = BigDecimal.ZERO;
        }
        if (this.sgstAmount == null) {
            this.sgstAmount = BigDecimal.ZERO;
        }
        if (this.igstAmount == null) {
            this.igstAmount = BigDecimal.ZERO;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public void addItem(PurchaseItem item) {
        items.add(item);
        item.setPurchase(this);
    }

    public void clearItems() {
        items.forEach(item -> item.setPurchase(null));
        items.clear();
    }
}
